### PR TITLE
`SpendPermissions` unit tests

### PR DESCRIPTION
**Pull request by amiecorso**
Unit tests for `SpendPermissions` contract

### CODE REVIEW

File: .gitignore
Comment (amiecorso) [inline .gitignore:21]: do you mind if we include this in the .gitignore so I can mess with my settings.json?

File: test/src/SpendPermissions/getCurrentCycle.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/getCurrentCycle.t.sol:18]: there's a final conditional branch when setting `uint48 end` to deal with overflows that would be good to check explicitly

File: test/src/SpendPermissions/getHash.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/getHash.t.sol:15]: Would add cases to check for replay conditions. Does the same recurring allowance have the same hash if we switch contracts or chains?

File: test/src/SpendPermissions/isAuthorized.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/isAuthorized.t.sol:15]: we generally want a test for every branch of logic, not just outcome. I think it will become more obvious which case is missing if you reword `true/false` to describe what preconditions are met to form that outcome. Hint, there are two cases where `isAuthorized` returns `false`

File: test/src/SpendPermissions/permit.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/permit.t.sol:16]: same thins as `approve` for when there are desired effects, we want tests to make sure those effects happen
- `test_approve_success_isAuthorized`
- `test_approve_success_emitsEvent`

File: test/src/SpendPermissions/revoke.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/revoke.t.sol:16]: ditto on tests to describe the effects in success case
- `test_revoke_success_isNotAuthorized`
- `test_revoke_success_emitsEvent`

File: test/src/SpendPermissions/useRecurringAllowance.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/useRecurringAllowance.t.sol:18]: typo 
```suggestion
    function test_useRecurringAllowance_revert_withdrawValueOverflow() public {}
```

File: test/src/SpendPermissions/useRecurringAllowance.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/useRecurringAllowance.t.sol:23]: Main logical branch here worth confirming is ability to work on both ETH and ERC20s.

We also want tests to describe the effects in success case
- `test_useRecurringAllowance_success_updatesState`
- `test_useRecurringAllowance_success_emitsEvent`

File: test/src/SpendPermissions/withdraw.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/withdraw.t.sol:19]: will leave it to you to think through more of the desired effects we need to check!

**Review by ilikesymmetry — COMMENTED** [review]
Good first pass! Think we'll be ready to start soon after getting a good lay of the land

File: test/src/SpendPermissions/getCurrentCycle.t.sol
Comment (ilikesymmetry) [inline test/src/SpendPermissions/getCurrentCycle.t.sol:154]: This seems fine to me 👍

File: test/src/SpendPermissions/useRecurringAllowance.t.sol
Comment (amiecorso) [reply test/src/SpendPermissions/useRecurringAllowance.t.sol]: since `_useRecurringAllowance` doesn't itself make the `_transferFrom`  call (and thus succeeds trivially regardless of the value for `token` in the recurring allowance struct), I'm going to test the ETH v.s. token as part of the `withdraw` test paths

**Comment by ilikesymmetry** [comment]
going to merge for now so I can continue developing on top of this for Wilson's changes in #70! Will drop comments tomorrow to address in a follow up PR if needed

**Review by ilikesymmetry — APPROVED** [review]

**Review by Nkuba2902 — APPROVED** [review]

**Pull request merged**
